lib/zclient.c

```diff
@@ -1461,6 +1462,223 @@ zebra_interface_vrf_update_read (struct stream *s, vrf_id_t vrf_id,
+ * immediately reads the answer from the input buffer.
+ *
+ * @param zclient Zclient used to connect to label manager (zebra)
+ * @result Result of response
+ */
+int
+lm_label_manager_connect (struct zclient *zclient)
+{
+  int ret;
+  struct stream *s;
+  u_char result;
+
+  zlog_debug ("Connecting to Label Manager");
+  if (zclient->sock < 0)
+    return -1;
+
+  /* send request */
+  s = zclient->obuf;
+  stream_reset (s);
+  zclient_create_header (s, ZEBRA_LABEL_MANAGER_CONNECT, VRF_DEFAULT);
+
+  /* proto */
+  stream_putc (s, zclient->redist_default);
+  /* instance */
+  stream_putw (s, zclient->instance);
+
+  /* Put length at the first point of the stream. */
+  stream_putw_at(s, 0, stream_get_endp(s));
+
+  ret = writen (zclient->sock, s->data, stream_get_endp (s));
+  if (ret < 0)
+    {
+      zlog_err ("%s: can't write to zclient->sock", __func__);
+      close (zclient->sock);
+      zclient->sock = -1;
+      return -1;
+    }
+  if (ret == 0)
+    {
+      zlog_err ("%s: zclient->sock connection closed", __func__);
+      close (zclient->sock);
+      zclient->sock = -1;
+      return -1;
+    }
+  zlog_debug ("%s: Label manager connect request (%d bytes) sent", __func__, ret);
+
+  /* read response */
+  s = zclient->ibuf;
+  stream_reset (s);
+
+  u_int16_t size;
+  u_char marker;
+  u_char version;
+  vrf_id_t vrf_id;
+  u_int16_t cmd;
+  ret = zclient_read_header (s, zclient->sock, &size, &marker, &version,
+                             &vrf_id, &cmd);
+  if (ret != 0 || cmd != ZEBRA_LABEL_MANAGER_CONNECT) {
+      zlog_err ("%s: Invalid Label Manager Connect Message Reply Header", __func__);
+      return -1;
+  }
+  /* result */
+  result = stream_getc(s);
+  zlog_debug ("%s: Label Manager connect response (%d bytes) received, result %u",
+              __func__, size, result);
+
+  return (int)result;
+}
+
+/**
+ * Function to request a label chunk in a syncronous way
+ *
+ * It first writes the request to zlcient output buffer and then
+ * immediately reads the answer from the input buffer.
+ *
+ * @param zclient Zclient used to connect to label manager (zebra)
+ * @param keep Avoid garbage collection
+ * @param chunk_size Amount of labels requested
+ * @param start To write first assigned chunk label to
+ * @param end To write last assigned chunk label to
+ * @result 0 on success, -1 otherwise
+ */
+int
+lm_get_label_chunk (struct zclient *zclient, u_char keep, uint32_t chunk_size,
+                    uint32_t *start, uint32_t *end)
+{
+  int ret;
+  struct stream *s;
+
+  zlog_debug ("Getting Label Chunk");
+  if (zclient->sock < 0)
+    return -1;
+
+  /* send request */
+  s = zclient->obuf;
+  stream_reset (s);
+  zclient_create_header (s, ZEBRA_GET_LABEL_CHUNK, VRF_DEFAULT);
+  /* keep */
+  stream_putc (s, keep);
+  /* chunk size */
+  stream_putl (s, chunk_size);
+  /* Put length at the first point of the stream. */
+  stream_putw_at(s, 0, stream_get_endp(s));
+
```

Comment:
Hm, this new requirement changes things quite a bit, I'd say that the label allocation algorithm will start to get "interesting" now. To simplify the code and improve performance, we might
consider the idea of using interval trees instead of linked lists to store the label chunks.